compiler/rustc_const_eval/src/interpret/machine.rs

```diff
@@ -484,13 +497,29 @@ pub macro compile_time_machine(<$mir: lifetime, $tcx: lifetime>) {
         Pointer::new(None, Size::from_bytes(addr))
     }
 
+    #[inline(always)]
+    fn ptr_from_casted_addr(
+        _ecx: &InterpCx<$mir, $tcx, Self>,
+        addr: u64,
+    ) -> Pointer<Option<AllocId>> {
+        Pointer::new(None, Size::from_bytes(addr))
+    }
+
+    #[inline(always)]
+    fn expose_ptr(
+        _ecx: &mut InterpCx<$mir, $tcx, Self>,
+        _ptr: Pointer<AllocId>,
+    ) -> InterpResult<$tcx> {
+        Ok(())
```

Comment:
This should throw an error, we don't allow these casts in CTFE. (Same for ptr_from_casted_addr.)

The way to throw an error differs between the CTFE interpreter and ConstProp, so you probably have to remove the functions here and instead add them to the various `Machine` trait impls.